### PR TITLE
Add an Epoll benchmark

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
Motivation:
Optimizing the Epoll channel needs an objective measure of how fast
it is.

Modification:
Add a simple, closed loop,  ping-pong benchmark.

Result:
Benchmark can be used to measure #7816

### CODE REVIEW

**Comment by carl-mastrangelo** [comment]
```
Result "io.netty.microbench.channel.epoll.EpollSocketChannelBenchmark.pingPong":
  22614.403 ±(99.9%) 797.263 ops/s [Average]
  (min, avg, max) = (21093.160, 22614.403, 24977.387), stdev = 918.130
  CI (99.9%): [21817.140, 23411.666] (assumes normal distribution)


# Run complete. Total time: 00:00:41

Benchmark                              Mode  Cnt      Score     Error  Units
EpollSocketChannelBenchmark.pingPong  thrpt   20  22614.403 ± 797.263  ops/s
```

File: microbench/src/main/java/io/netty/microbench/channel/epoll/EpollSocketChannelBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+public class EpollSocketChannelBenchmark extends AbstractMicrobenchmark {
+
+  private EpollEventLoopGroup group;
+  private Channel serverChan;
+  private Channel chan;
+  private ByteBuf abyte;
+  private ScheduledFuture<?> future;
+
+  @Setup
+  public void setup() throws Exception {
```
Comment (normanmaurer) [inline microbench/src/main/java/io/netty/microbench/channel/epoll/EpollSocketChannelBenchmark.java]: please use 4 spaces

File: microbench/src/main/java/io/netty/microbench/channel/epoll/EpollSocketChannelBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        .group(group)
+        .childHandler(new ChannelInitializer<Channel>() {
+          @Override
+          protected void initChannel(Channel ch) {
+            ch.pipeline().addLast(new ChannelDuplexHandler() {
+              @Override
+              public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                if (msg instanceof ByteBuf) {
+                  ctx.writeAndFlush(msg, ctx.voidPromise());
+                  return;
```
Comment (normanmaurer) [inline microbench/src/main/java/io/netty/microbench/channel/epoll/EpollSocketChannelBenchmark.java]: return is not needed.

File: microbench/src/main/java/io/netty/microbench/channel/epoll/EpollSocketChannelBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        .sync()
+        .channel();
+    chan = new Bootstrap()
+        .channel(EpollSocketChannel.class)
+        .handler(new ChannelInitializer<Channel>() {
+          @Override
+          protected void initChannel(Channel ch) {
+            ch.pipeline().addLast(new ChannelDuplexHandler() {
+
+              ChannelPromise lastWritePromise;
```
Comment (normanmaurer) [inline microbench/src/main/java/io/netty/microbench/channel/epoll/EpollSocketChannelBenchmark.java]: private ?

File: microbench/pom.xml [hunk 4]
```diff
@@ -78,6 +78,11 @@
       <artifactId>netty-codec-redis</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
+      <groupId>${project.groupId}</groupId>
+      <artifactId>netty-transport-native-epoll</artifactId>
+      <version>${project.version}</version>
```
Comment (normanmaurer) [inline microbench/pom.xml]: classifier is missing:

```xml
          <classifier>linux-x86_64</classifier>
```

Comment (carl-mastrangelo) [reply microbench/pom.xml]: Done.

**Comment by carl-mastrangelo** [comment]
@normanmaurer PTAL

**Comment by normanmaurer** [comment]
@carl-mastrangelo thanks! Made a small change so we are able to compile the microbenchmarks still on macOS and pulled in.

**Comment by normanmaurer** [comment]
Cherry-picked as 379a56ca49195396bc78f1d96ec2cd7164255b96

**Pull request closed without merging**